src/metabase/pulse/markdown.clj

```diff
@@ -0,0 +1,320 @@
+      ;; Add a soft hyphen around certain chars to avoid triggering formatting in Slack
+      (str/replace "&" "\u00ad&\u00ad")
+      (str/replace ">" "\u00ad>\u00ad")
+      (str/replace "<" "\u00ad<\u00ad")
+      (str/replace "*" "\u00ad*\u00ad")
+      (str/replace "_" "\u00ad_\u00ad")
+      (str/replace "`" "\u00ad`\u00ad")
+      (str/replace "~" "\u00ad~\u00ad")))
+
+(defn- resolve-uri
+  "If the provided URI is a relative path, resolve it relative to the site URL so that links work
+  correctly in Slack/Email."
+  [uri]
+  (when uri
+    (if-let [site-url (public-settings/site-url)]
+      (.toString (.resolve (new URI ^String site-url) ^String uri))
+      uri)))
+
+(defn- ast->mrkdwn
+  "Takes an AST representing Markdown input, and converts it to a mrkdwn string that will render nicely in Slack.
+
+  Some of the differences to Markdown include:
+    * All headers are just rendered as bold text.
+    * Ordered and unordered lists are printed in plain text.
+    * Inline images are rendered as text that links to the image source, e.g. <image.png|[Image: alt-text]>."
+  [{:keys [tag attrs content]}]
+  (let [resolved-content (if (string? content)
+                           (escape-text content)
+                           (map #(if (string? %)
+                                   (escape-text %)
+                                   (ast->mrkdwn %))
+                                content))
+        joined-content   (str/join resolved-content)]
+    (case tag
+      :document
+      joined-content
+
+      :paragraph
+      (str joined-content "\n")
+
+      :soft-line-break
+      " "
+
+      :hard-line-break
+      "\n"
+
+      (:heading :bold)
+      (str "*" joined-content "*")
+
+      :italic
+      (str "_" joined-content "_")
+
+      :code
+      (str "`" joined-content "`")
+
+      :codeblock
+      (str "```\n" joined-content "```")
+
+      :blockquote
+      (let [lines (str/split-lines joined-content)]
+        (str/join "\n" (map #(str ">" %) lines)))
+
+      :link
+      (let [resolved-uri (resolve-uri (:href attrs))]
+        (if (contains? #{:image :image-ref} (:tag (first content)))
+          ;; If this is a linked image, add link target on separate line after image placeholder
+          (str joined-content "\n(" resolved-uri ")")
+          (str "<" resolved-uri "|" joined-content ">")))
+
+      :link-ref
+      (if-let [resolved-uri (resolve-uri (-> attrs :reference :attrs :url))]
+        (str "<" resolved-uri "|" joined-content ">")
+        joined-content)
+
+      :auto-link
+      (str "<" (:href attrs) ">")
+
+      :mail-link
+      (str "<" (:address attrs) ">")
+
+      ;; list items might have nested lists or other elements, which should have their indentation level increased
+      (:unordered-list-item :ordered-list-item)
+      (let [indented-content (->> (rest resolved-content)
+                                  str/join
+                                  str/split-lines
```

Comment:
what's the purpose of the str/join str/split-lines sequence here?

Reply:
`resolved-content` contains a list of strings, but they correspond to subnodes in the AST, and might not already be split on newlines.

So we need to join, and then re-split the content on newlines so that we know where to add the indentations.